biosimulations/apps/dispatch-api/src/app/app.controller.ts

```diff
@@ -138,102 +127,66 @@ export class AppController implements OnApplicationBootstrap {
     };
   }
 
-  @Get('result/:uuid')
+  @Get('result/structure/:uuid')
   @ApiResponse({
     status: 200,
-    description: 'Get Simulation Results',
+    description: 'Get results structure (SEDMLS and TASKS)',
     type: Object,
   })
-  getVisualizationData(
-    @Param('uuid') uId: string,
-    @Query('chart') chart: boolean
-  ) {
-    const jsonResults: Dic = {};
+  async getResultStructure(@Param('uuid') uId: string) {
     const fileStorage = process.env.FILE_STORAGE || '';
+    const structure: any = {};
 
-    const csvFilePath = path.join(fileStorage, 'simulations', uId, 'out');
-
-    const directoryList = fs.readdirSync(csvFilePath);
-    // NOTE: job.output is the Log file generated by the SBATCH simulation job
-    const logFileIndex = directoryList.indexOf('job.output');
-    directoryList.splice(logFileIndex);
-
-    directoryList.forEach((directoryName: string) => {
-      const fileList = fs.readdirSync(path.join(csvFilePath, directoryName));
+    const resultPath = path.join(fileStorage, 'simulations', uId, 'out');
+    const sedmls = fs.readdirSync(resultPath);
+    // Removing log file names 'job.output'
+    sedmls.splice(sedmls.indexOf('job.output'), 1);
 
-      fileList.forEach(async (filename: string) => {
-        if (filename.endsWith('csv')) {
-          const taskName = filename.split('.csv')[0];
-          const filePath = path.join(csvFilePath, directoryName, filename);
-          this.logger.log('Reading file: ' + filePath);
-
-          if (jsonResults[directoryName] === undefined) {
-            jsonResults[directoryName] = {};
-          }
-
-          const jsonPath = filePath.split('.csv')[0] + '.json';
-
-          const jsonResult: object[] = [];
-          setTimeout(() => {
-            fs.createReadStream(filePath)
-              .pipe(
-                csv2Json.default({
-                  separator: ',',
-                })
-              )
-              .pipe(fs.createWriteStream(jsonPath));
-          }, 0);
-
-          const fileContent = fs.readFileSync(jsonPath);
-          jsonResults[directoryName][taskName] = JSON.parse(
-            fileContent.toString()
-          );
+    sedmls.forEach((sedml) => {
+      structure[sedml] = [];
+      const taskFiles = fs.readdirSync(path.join(resultPath, sedml));
```

Comment:
Could you change these to use an async method? The sync methods will only allow the API to  handle one client at a time

Reply:
I can, but by default the async method for file read expects a callback which is fine for a non API use case, we need to return the data as JSON in response which by itself is a synchronous use case in my opinion, do you want me to wrap the async call in a promise and then return that promise? This is the only way that I think which we can make it async.

But internally it would actually by synchronous because NestJS will wait for the promise to get resolved before returning the response, or that won't be the case and I'm missing something here?

Reply:
So while nest will resolve the promise, it will be able to handle other user requests. 
Here is a look at how to do this using promised. Essentially just using the await keyword will have node do all the work for you . 
https://nodejs.org/api/fs.html#fs_promise_example

Reply:
Done